### PR TITLE
Generic middleware stack

### PR DESCRIPTION
**Pull request by georgegg**
## Which problem is this PR solving?

Framework has no generic middleware stack mechanism for default http component.
(issue #47)

## Short description of the changes

This PR addresses issue #47 and introduces a generic middleware stack that can be setup in main service as an OptionFunc and then in the default http component as a slice of MiddlewareFunc, that are applied on all routes chained in order after the default middlewares of auth and tracing.

### CODE REVIEW

**Comment by mantzas** [comment]
@sotirispl @cpapidas in order to implement some needed middleware's (for CORS) which is blocking for the config mgmt working group the PR comes before the actual ticket. Wait for the ticket to understand the what ans see then the how. /cc @georgegg

File: option.go [hunk 1]
```diff
@@ -22,6 +22,18 @@ func Routes(rr []http.Route) OptionFunc {
 	}
 }
 
+// Middlewares option for adding generic middlewares to the default HTTP component.
+func Middlewares(mm []http.MiddlewareFunc) OptionFunc {
```
Comment (mantzas) [inline option.go]: i propose to turn this into a variadic function

File: option.go [hunk 2]
```diff
@@ -22,6 +22,18 @@ func Routes(rr []http.Route) OptionFunc {
...
+	return func(s *Service) error {
+		if len(mm) == 0 {
+			return nil
```
Comment (mantzas) [inline option.go]: if the user supplies middlewares via the option he should provide at least one. this should be a actual error.

File: option_test.go [hunk 3]
```diff
@@ -1,6 +1,7 @@
 package patron
 
 import (
+	http2 "net/http"
```
Comment (mantzas) [inline option_test.go]: when there is a package name conflict "http" the std package one stays the same and the custom package has to be aliased

File: option_test.go [hunk 4]
```diff
@@ -34,6 +35,37 @@ func TestRoutes(t *testing.T) {
...
+	}
+	type args struct {
+		mm []http.MiddlewareFunc
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{"success", args{mm: []http.MiddlewareFunc{middleware}}, false},
```
Comment (mantzas) [inline option_test.go]: the error check is missing

File: service_test.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package patron
 import (
 	"context"
 	"math/rand"
+	http2 "net/http"
```
Comment (mantzas) [inline service_test.go]: package naming like before, use patronhttp as the http package alias for patron

File: service_test.go [hunk 6]
```diff
@@ -14,22 +15,28 @@ import (
...
 	tests := []struct {
 		name    string
 		args    args
 		wantErr bool
 	}{
-		{"success", args{name: "test", opt: Routes([]http.Route{route})}, false},
-		{"failed missing name", args{name: "", opt: Routes([]http.Route{route})}, true},
-		{"failed missing routes", args{name: "test", opt: Routes([]http.Route{})}, true},
+		{"success", args{name: "test", opt: []OptionFunc{Routes([]http.Route{route}), Middlewares([]http.MiddlewareFunc{middleware})}}, false},
+		{"success empty middlewares", args{name: "test", opt: []OptionFunc{Routes([]http.Route{route}), Middlewares([]http.MiddlewareFunc{})}}, false},
```
Comment (mantzas) [inline service_test.go]: failed missing middlewares

**Review by mantzas — CHANGES_REQUESTED** [review]

File: option.go (hunk 1 above)
Comment (georgegg) [reply option.go]: I followed the Routes func signature (above) and I wanted to make them both variadic but afraid to break compatibility :/ 
What do you think?

Comment (mantzas) [reply option.go]: I dont see how you break compatibility here?
i propose this signature on the new option:
```go
func Middlewares(mm ...http.MiddlewareFunc) OptionFunc {
```

**Review by mantzas — APPROVED** [review]

**Comment by cpapidas** [comment]
Could you please add the "how to use it" information in the readme.md file and maybe add a dummy middleware in the first example?

**Review by sotirispl — APPROVED** [review]

**Review by cpapidas — APPROVED** [review]

**Pull request merged**